README.rst

```diff
@@ -46,10 +47,69 @@ ease installation.  To use in edx-platform, do the following:
 override the settings in `EDXAPP_ENV_EXTRA` instead.
 
-Documentation
--------------
-
-The full documentation is at https://openedx-completion-aggregator.readthedocs.org.
+Design
+------
+
+The completion aggregator is designed to facilitate working with course-level,
+chapter-level, and other aggregated percentages of course completion as
+represented by the BlockCompletion model (from the edx-completion djangoapp).
+By storing these values in the database, we are able to quickly return
+information for all users in a course.
+
+Each type of XBlock (or XModule) is assigned a completion mode of
+"Completable", "Aggregator", or "Excluded".
+
+A "completable" block is one that can directly be completed, either by viewing it
+on the screen, by submitting a response, or by some custom defined means.  When
+completed by, a BlockCompletion is created for that user with a value of 1.0
+(any value between 0.0 and 1.0 is allowed).  Completable blocks always have a
+maximum possible value of 1.0.
+
+An "excluded" block is ignored for the purposes of completion.  It always has
+a completion value of 0.0, and a maximum possible value of 0.0.  If an excluded
+block has children, those are also ignored for the purposes of completion.
+
+An "aggregator" block is one that contains other blocks.  It cannot be directly
+completed, but has an aggregate completion value equal to the sum of the
+completion values of its immediate children, and a maximum possible value equal
+to the sum of the maximum possible values of its immediate children (1.0 for
+completable blocks, 0.0 for excluded blocks, and the calculated maximum for any
+contained aggregators).  If an aggregator has a maximum possible value of 0.0,
+(either it has no children, or all its children are excluded), it is always
+considered complete.
+
+To calculate aggregations for a user, the course graph is retrieved from the
+modulestore (using block transformers) to determine which blocks are contained
+by each aggregator, and values are summed recursively from the course block on
+down.  Values for every node in the whole tree can be calculated in a single
+traversal.  These calculations can either be performed "read-only" (to get the
+latest data for each user), or "read-write" to store that data in the
+completion_aggregator.Aggregator model.
+
+During regular course interaction, a learner will calculate aggregations on the
+fly to get the latest information.  However, on-the-fly calculations are too
+expensive when performed for all users in a course, so periodically (every hour
+or less), a task is run to to calculate all aggregators that have gone out of
+date in the previous hour, and store those values in the database.  These
+stored values are then used for reporting on course-wide completion (for course
+admin views).
+
+By tracking which blocks have been changed recently (in the StaleCompletion
+table) These stored values can also be used to shortcut calculations for
```

Comment:
Maybe this sentence can be different:
"Recent changes are tracked using the `StaleCompletion` model. The values stored in the `StaleCompletion` table can also be used to shortcut calculations for..."